File: ts-sdk/src/health/run.health.check.ts
import http from "http";
import url from "url";
import { HealthCheck } from "../handlers";
import { MetricsManager } from "../metrics";
import { Logger, assertExists } from "../utils";

export type RunHealthCheck = (handler?: HealthCheck) => Promise<void>;

export function provideRunHealthCheck(
  healthCheckPort: number,
  metricsManager: MetricsManager,
  logger: Logger
): RunHealthCheck {
  assertExists(metricsManager, "metricsManager");
  assertExists(logger, "logger");

  return async function runHealthCheck(handler?: HealthCheck) {
    const server = http.createServer();
    server.on("request", async (req, res) => {
      const path = url.parse(req.url!).pathname;
      if (path == "/health") {
        res.statusCode = 200;
        let errors: string[] = [];
        try {
          if (handler) {
            const response = await handler();
            if (response && response.length > 0) {
              errors = response;
            }
          }
        } catch (e) {
          logger.error(`${new Date().toISOString()}    healthCheck`);
          logger.error(e);
          res.statusCode = 500;
          errors = [e.message];
        }
        res.end(
          JSON.stringify({
            errors,
            metrics: metricsManager.flushMetrics(),
          })
        );
      }
    });

    server.listen(healthCheckPort);
  };
}

File: py-sdk/src/forta_bot/health/run_health_check.py
from datetime import datetime
from aiohttp import web
from typing import Callable, Optional
from ..metrics import MetricsManager
from ..utils import assert_exists, Logger
from ..common import HealthCheck


RunHealthCheck = Callable[[Optional[HealthCheck]], None]


def provide_run_health_check(metrics_manager: MetricsManager, health_check_port: int, logger: Logger):
    assert_exists(metrics_manager, 'metrics_manager')
    assert_exists(logger, 'logger')

    async def run_health_check(handler: Optional[HealthCheck] = None):
        # define the HTTP request handler
        async def health_check_handler(request):
            status: int = 200
            errors: list[str] = []
            try:
                if handler:
                    response = await handler()
                    if response and len(response) > 0:
                        errors = response
            except Exception as e:
                logger.error(f'{datetime.now().isoformat()}    health_check')
                logger.error(e)
                status = 500
                errors = [str(e)]
            return web.json_response({'errors': errors, 'metrics': metrics_manager.flush_metrics()}, status=status)

        # run the http server
        server = web.Application()
        server.add_routes([web.get('/health', health_check_handler)])
        runner = web.AppRunner(server)
        await runner.setup()
        site = web.TCPSite(runner, port=health_check_port)
        await site.start()

    return run_health_check
